exporter/kinesisexporter/exporter.go

```diff
@@ -57,32 +69,29 @@ func (e *Exporter) Shutdown(_ context.Context) error {
 }
 
 func (e *Exporter) pushTraces(_ context.Context, td pdata.Traces) (int, error) {
-	pBatches, err := jaegertranslator.InternalTracesToJaegerProto(td)
+	data, err := e.marshaller.MarshalTraces(td)
+	var exportErr []error
 	if err != nil {
 		e.logger.Error("error translating span batch", zap.Error(err))
-		return td.SpanCount(), consumererror.Permanent(err)
-	}
+		exportErr = append(exportErr, consumererror.Permanent(err))
 
-	// TODO: Use a multi error type
-	var exportErr error
-	for _, pBatch := range pBatches {
-		for _, span := range pBatch.GetSpans() {
-			if span.Process == nil {
-				span.Process = pBatch.Process
-			}
-
-			spanBytes, err := span.Marshal()
-			if err != nil {
-				e.logger.Error("error marshaling span to bytes", zap.Error(err))
-				return td.SpanCount(), consumererror.Permanent(err)
-			}
+		if data == nil {
+			return td.SpanCount(), err
+		}
+	}
 
-			if err = e.producer.put(spanBytes, span.SpanID.String()); err != nil {
+	var droppedSpans int
+	for _, msg := range data {
+		if msg.Value != nil {
+			if err = e.producer.put(msg.Value, uuid.New().String()); err != nil {
```

Comment:
Isn't this a big breaking change? If I understand correctly, we are not pushing spans to kinesis with a random routing/sharding key. Earlier we guaranteed that spans from the same trace would end up on the same kinesis shard (unless there is a partition). This meant each kinesis consumer would be guaranteed to receive all spans for any given trace. With this change, spans from a single trace may or may not end up being processed by the same consumer.

Reply:
Looking at previous version, it looks like we were using Span ID as the key. I remember the original implementation used the Trace ID. Not sure when or why it was changed. If it has been like this for a few releases, then it probably is not that big of a deal but still it is easy to imaging a trace processing pipeline expecting trace ID based sharding.

Reply:
Looks like it slipped past the review in last PR: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1345/files#diff-8127d5206d01eb17a91df0f737f9e9f950294c0b473557f1177cf49f5fd57000R80

Here we are using span ID as the partition key but the function we replaced is actually using trace ID internally:

https://github.com/signalfx/opencensus-go-exporter-kinesis/blob/master/kinesis.go#L258 and https://github.com/signalfx/opencensus-go-exporter-kinesis/blob/master/shardProducer.go#L89

Looks like the partitioning logic from opencensus-go-exporter-kinesis was not replicated in the replacement otel exporter.

Reply:
This uncovers more breaking changes. For example, messages no longer start with the magic byte anymore and messages aren't gzipped. https://github.com/signalfx/opencensus-go-exporter-kinesis/blob/master/shardProducer.go#L23 

Looks like in order to not break things, we'd need to implement at least three features in this exporter:

- Use trace ID as partition key. Means batches can't be published as they come and probably need to be re-batched by kinesis partitioning scheme as dictated by the trace ID.
- Gzip the messages before publishing to kinesis. This reduces the overall size anywhere between 4-10x if my memory serves right.
- Prepend the magic byte header to messages. The header should be the same as the one used by opencensus exporter when exporting in Jaeger. OTLP should use a different header. It is used by consumers to determine the encoding of the messages and whether they are gzipped or not.

Reply:
I think the easiest way to retain everything and not break existing users would be copy over github.com/signalfx/opencensus-go-exporter-kinesis as it and then replace the OpenCensus exporter functionality with OpenTelemetry and just submit that as a PR. Then rebase this PR on once that one is merged. Since it is pre-existing code that the collector was already using, I don't think we'll need to fix or improve things in code that gets copied over. 

An even easier path would be to fork that repo and make it an OpenTelemetry exporter. Then use it from this exporter as a dependency.

Reply:
Thanks for the comments, definitely big oversight on our part with the differences between omnition and opencensus.

Our original intention with using the omnition producer was that we could leverage byte publishing without exporting a specific data format. I think this is an interface we would want for the kinesis producer since kinesis itself accepts any byte data. The features you previously mentioned are definitely desired. 

As an alternative to your suggestion, I did have a solution in mind for each of the features you mentioned.

> Use trace ID as partition key. Means batches can't be published as they come and probably need to be re-batched by kinesis partitioning scheme as dictated by the trace ID.
We can append a partition key field to the [Message struct](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1403/files#diff-24959c12e0e39c75b8b1221b12eaa1ec17afbdeb57d6513eba49cbc2d85a4034R30), this is populated during marshaling and so we can use the trace ID for the jaeger marshaller 

> Gzip the messages before publishing to kinesis. This reduces the overall size anywhere between 4-10x if my memory serves right.
This would be a generally desired feature for all payloads and could be part of the marshalling step.

> Prepend the magic byte header to messages. The header should be the same as the one used by opencensus exporter when exporting in Jaeger. OTLP should use a different header. It is used by consumers to determine the encoding of the messages and whether they are gzipped or not.
Can be added as a function to the jaeger and OT marshaller.

The downside here is it'll be a duplication of the work done in the opencensus-exporter with the upside of maintaining a byte publishing interface.

I'm uncertain of the required changes to make the opencensus-exporter compatible with what we have. Could you break down what work would need to be done to make it a OpenTelemetry exporter? If it is the easier path then we will go down that path

Also as a heads up we have some internal deadlines coming up, is it alright if we pick this work up after Monday 2nd October (US). I do realise that there are breaking changes right now, we would be fine with reverting #1345 as a temporary solution for backwards compatibility.

Reply:
No worries and no rush. Let's discuss once you are back. In the meanwhile, can we revert the merged PR so it doesn't make it into a release and break tracing pipelines for existing users?

Reply:
Based on the discussion I assume this PR is not yet ready for the review? If so please convert it to draft.

Reply:
This is not ready and will probably need to be redone or rebased after another (not yet created) PR is merged. Will close for now to clean up the board.